--- a/src/kianv_harris_edition/load_decoder.v
+++ b/src/kianv_harris_edition/load_decoder.v
@@ -40,7 +40,10 @@
       case (1'b1)
         is_lb:  LOADop = `LOAD_OP_LB;
         is_lbu: LOADop = `LOAD_OP_LBU;
-        is_lhu: LOADop = `LOAD_OP_LHU;
+        is_lhu: begin
+          LOADop = `LOAD_OP_LHU;
+          is_load_unaligned = addr_align_bits[0];
+        end
 
         is_lh: begin
           LOADop = `LOAD_OP_LH;
