--- a/src/kianv_harris_edition/datapath_unit.v
+++ b/src/kianv_harris_edition/datapath_unit.v
@@ -217,14 +217,26 @@
       .q     (amo_buffer_addr_value)
   );
 
+  wire amo_reserved;
   dff_kianV #(1) amo_reserved_state_load_I (
       .resetn(resetn),
       .clk   (clk),
       .d     (amo_buffered_data),
       .en    (amo_set_reserved_state_load),
-      .q     (amo_reserved_state_load)
+      .q     (amo_reserved)
   );
 
+  wire [31:0] amo_reserved_addr;
+  dff_kianV #(32) amo_reserved_addr_I (
+      .resetn(resetn),
+      .clk   (clk),
+      .d     (amo_buffer_addr_value),
+      .en    (amo_set_reserved_state_load && amo_buffered_data),
+      .q     (amo_reserved_addr)
+  );
+
+  assign amo_reserved_state_load = amo_reserved && amo_reserved_addr == ALUResult;
+
   dff_kianV #(32, RESET_ADDR) OldPC_I (
       resetn,
       clk,
--- a/src/kianv_harris_edition/main_fsm.v
+++ b/src/kianv_harris_edition/main_fsm.v
@@ -729,10 +729,11 @@
       end
 
       AMO_ADDR: begin
-        ALUSrcA              = `SRCA_RD1_BUF;
-        ALUSrcB              = `SRCB_CONST_0;
-        ALUOp                = `ALU_OP_ADD;
-        amo_buffered_address = 1'b1;
+        ALUSrcA                     = `SRCA_RD1_BUF;
+        ALUSrcB                     = `SRCB_CONST_0;
+        ALUOp                       = `ALU_OP_ADD;
+        amo_buffered_address        = 1'b1;
+        amo_set_reserved_state_load = is_amo_sc_w;
       end
 
       AMO_LR_READ: begin
