--- a/src/kianv_harris_edition/main_fsm.v
+++ b/src/kianv_harris_edition/main_fsm.v
@@ -826,7 +826,7 @@
 
       CSR_FAULT_0: begin
         cause           = `EXC_ILLEGAL_INSTRUCTION;
-        badaddr         = {25'b0, op};
+        badaddr         = {funct7, Rs2, Rs1, funct3, Rd, op};
         exception_event = 1'b1;
       end
 
@@ -875,7 +875,7 @@
 
       ILLEGAL_0: begin
         cause           = `EXC_ILLEGAL_INSTRUCTION;
-        badaddr         = {25'b0, op};
+        badaddr         = {funct7, Rs2, Rs1, funct3, Rd, op};
         exception_event = 1'b1;
       end
 
